File: Data Structures & Algorithms/eating-bananas/submission-1.py
class Solution:
    def ceil(self, numerator: int, denominator: int) -> int:
        return -(-numerator // denominator)

    def minEatingSpeed(self, piles: List[int], h: int) -> int:
        highest = max(piles)
        k = 0

        low = 1
        high = highest
        while low <= high:
            mid = low + ((high-low) // 2)
            time = sum([self.ceil(pile,mid) for pile in piles])
            if time > h:
                low = mid+1
            else:
                k = mid
                high = mid-1
        return k

File: Data Structures & Algorithms/search-2d-matrix/submission-1.cpp
class Solution {
public:
    bool searchMatrix(vector<vector<int>>& matrix, int target) {
        int n = matrix.size(), m = matrix[0].size();
        int low = 0, high = n*m-1;

        while (low <= high){
            int middle = low + (high-low) / 2;
            int value = matrix[middle / m][middle % m];
            
            if (value == target) return true;
            if (value < target) low = middle+1;
            else high = middle-1;
        }

        return false;
    }
};
